js/ui/light.js

```diff
@@ -0,0 +1,256 @@
+            color = 'color' in options ? parseColor(options.color) : startColor,
+            direction = 'direction' in options ? options.direction : startDirection,
+            intensity = 'intensity' in options ? +options.intensity : startIntensity;
+
+        // light.anchor does not transition.
+        if (anchor !== startAnchor) this._setLightAnchor(anchor);
+
+        this.activeLightTransitions.color = color !== startColor;
+        this.activeLightTransitions.direction = direction.some(function(c, i) { return c !== startDirection[i]; });
+        this.activeLightTransitions.intensity = intensity !== startIntensity;
+
+        this.fire('lightstart', eventData);
+
+        clearTimeout(this._onLightEnd);
+
+        if (this.activeLightTransitions.direction) {
+            // Normalize new azimuthal [0, 360) coordinates
+            while (direction[1] >= 360) { direction[1] -= 360; }
+            while (direction[1] < 0)    { direction[1] += 360; }
+            // Clamp polar coordinates to [0, 180]
+            if (direction[2] > 180)     { direction[2] = 180; }
+            if (direction[2] < 0)       { direction[2] = 0; }
+
+            // Wrap prior azimuthal coordinate for use in interpolation
+            if (Math.abs(startDirection[1] - direction[1]) > 180) {
+                if (direction[1] >= 180) {
+                    startDirection[1] += 360;
+                } else {
+                    startDirection[1] -= 360;
+                }
+            }
+        }
+
+        this._ease(function(k) {
+            if (this.activeLightTransitions.color) {
+                this._setLightColor({
+                    rawColor: options.color,
+                    color: interpolate.color(startColor, color, k)
+                });
+            }
+            if (this.activeLightTransitions.direction) {
+                this._setLightDirection(interpolate.array(startDirection, direction, k));
+            }
+            if (this.activeLightTransitions.intensity) {
+                this._setLightIntensity(interpolate(startIntensity, intensity, k));
+            }
+            this._update();
+        }, function() {
+            if (options.delayEndEvents) {
+                this._onLightEnd = setTimeout(this._lightEnd.bind(this, eventData), options.delayEndEvents);
+            } else {
+                this._lightEnd(eventData);
+            }
+        }.bind(this), options);
+
+        return this;
+    },
+
+    _lightEnd: function(eventData) {
+        this.activeLightTransitions.color = false;
+        this.activeLightTransitions.direction = false;
+        this.activeLightTransitions.intensity = false;
+
+        this.fire('lightend', eventData);
+    },
+
+    _setLightAnchor: function(anchor) {
+        if (anchor === 'map' || anchor === 'viewport') {
+            this.painter.setLight({
+                anchor: anchor
+            });
+        } else throw new Error('light.anchor must be one of: `map`, `viewport`');
```

Comment:
`light` is associated with the style. 

We keep all [style validation code separate](https://github.com/mapbox/mapbox-gl-style-spec/tree/mb-pages/lib/validate) from GL JS so that we can have a standalone style validation utility. This is used, for example, by the style upload API. 

[Here is an example of where we use the style spec's validation code to validate layers](http://github.com/mapbox/mapbox-gl-js/blob/ece53f12f4bd72aaa555813a5363a1fd0c9d19f6/js/style/style.js#L497-L498)